frontend/lib/resolution/Resolver.cpp

```diff
@@ -6450,17 +6450,105 @@ resolveIterTypeWithTag(Resolver& rv,
   return yieldType;
 }
 
-struct ParamRangeInfo {
-  int64_t current;
-  int64_t end;
+struct BaseParamRangeInfo {
   const Type* yieldType = nullptr;
 
   // Optionally specified properties
   int64_t step = 1;
 
-  static optional<ParamRangeInfo> fromBound(Context* context, ResolutionResultByPostorderID& rr, const AstNode* node) {
-    ParamRangeInfo scratch;
+  virtual ~BaseParamRangeInfo() = default;
+
+  virtual bool done() const = 0;
+  virtual QualifiedType advance(Context* context) = 0;
+};
+
+template <typename R>
+static bool populateParamRangeInfoFromBound(R& scratch, optional<typename R::Iter> low, optional<typename R::Iter> hi, Range::OpKind opKind, int step, int numElts) {
```

Comment:
Could/should these two "populate" functions be part of a factory method on one of these structs?